sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -51,6 +51,11 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
 
   def reorder(plan: LogicalPlan, output: AttributeSet): LogicalPlan = {
     val (items, conditions) = extractInnerJoins(plan)
+    // Find the star schema joins. Currently, it returns the star join with the largest
+    // fact table. In the future, it can return more than one star join (e.g. F1-D1-D2
+    // and F2-D3-D4).
+    val starJoinPlans = StarSchemaDetection(conf).findStarJoins(items, conditions.toSeq)
```

Comment:
The dynamic programming solution proposed by Pat was also used in DB2. She is the mother of DB2. We knew the limit of this solution. It is unable to solve all the issues. [The above suggestion](https://github.com/apache/spark/pull/15363#issuecomment-285187051) by Ioana (who is the senior DB2 compiler expert) is in the right direction, IMO,